src/controller/CHIPDeviceController.cpp

```diff
@@ -1190,6 +1209,170 @@ void DeviceCommissioner::OnSessionEstablished()
+    ChipLogDetail(Controller, "Sent Attestation request, waiting for the Attestation Information");
+    return CHIP_NO_ERROR;
+}
+
+void DeviceCommissioner::OnAttestationFailureResponse(void * context, uint8_t status)
+{
+    ChipLogProgress(Controller, "Device failed to receive the Attestation Information Response: 0x%02x", status);
+    DeviceCommissioner * commissioner = reinterpret_cast<DeviceCommissioner *>(context);
+    commissioner->mAttestationResponseCallback.Cancel();
+    commissioner->mOnAttestationFailureCallback.Cancel();
+    // TODO: Map error status to correct error code
+    commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+}
+
+void DeviceCommissioner::OnAttestationResponse(void * context, chip::ByteSpan attestationElements, chip::ByteSpan signature)
+{
+    ChipLogProgress(Controller, "Received Attestation Information from the device");
+    DeviceCommissioner * commissioner = reinterpret_cast<DeviceCommissioner *>(context);
+
+    commissioner->mAttestationResponseCallback.Cancel();
+    commissioner->mOnAttestationFailureCallback.Cancel();
+
+    CHIP_ERROR err = commissioner->ValidateAttestationInfo(attestationElements, signature);
+    if (err != CHIP_NO_ERROR)
+    {
+        // Handle error, and notify session failure to the commissioner application.
+        ChipLogError(Controller, "Failed to validate the Attestation Information");
+        // TODO: Map error status to correct error code
+        commissioner->OnSessionEstablishmentError(CHIP_ERROR_INTERNAL);
+    }
+}
+
+CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(chip::ByteSpan attestationElements, chip::ByteSpan signature)
+{
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
+
+    Device * device = &mActiveDevices[mDeviceBeingPaired];
+
+    SetDeviceAttestationVerifier(Examples::GetExampleDACVerifier());
+    DeviceAttestationVerifier * dac_verifier = GetDeviceAttestationVerifier();
+
+    // Retrieve attestation challenge
+    PeerConnectionState * state = mSessionMgr->GetPeerConnectionState(
+        { mPairingSession.GetPeerNodeId(), mPairingSession.GetLocalKeyId(), mPairingSession.GetPeerKeyId(), mFabricIndex });
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+
+    VerifyOrReturnError(dac_verifier->VerifyAttestationInformation(
+                            attestationElements, state->GetSecureSession().GetAttestationChallenge(), signature, device->GetPAI(),
+                            device->GetDAC(), device->GetAttestationNonce()) == AttestationVerificationResult::kSuccess,
+                        CHIP_ERROR_INTERNAL);
+
+    // TODO: Step g/h validate CertDeclaration
+    // TODO: Step i: validate firmware information
+
+    ChipLogProgress(Controller, "Sending 'CSR request' command to the device.");
```

Comment:
Stringing this step here is the wrong place to do it. It should be done in OnAttestationResponse, but even then, we need to account for "inner" and "outer" state machine.

@cecille Is there something more that could be done here for the sequencing using the `AdvanceCommissioningStage()` method?

Reply:
Yeah, the eventual plan is to pull all the cert stuff into the state machine function (and possibly use a more sophistocated state machine). As always, the blocker is not technical, but time.

Reply:
will update to add new step to state machine.
will move CSR Request step into OnAttestationRespose as suggested.